### ISSUE TITLE
self-hosted-runner: detect non-GH runners that don't use `self-hosted`

### ISSUE THREAD

**Issue by woodruffw** [zizmorcore/zizmor#286]
The current audit assumes that `self-hosted` is a reliable sentinel for whether a job runs on a self-hosted runner or not, but that turns out to not be the case: self-hosted runners can seemingly use *any* labels, assuming the underlying runner controller/orchestrator (?) can handle them.

This can be seen in e.g. the [RunsOn](https://runs-on.com/runners/linux/) service, where they have routing keys like this for self-hosted runners:

```yaml
    runs-on:
      - runs-on=${{ github.run_id }}
      - runner=2cpu-linux-arm64
```

We should figure out a way to handle these. In practice there's probably nothing fully general here; we'll just need to walk the runs-on labels and flag anything that isn't a well-known label (which is a static list provided by GH itself).

**Comment by ubiratansoares** [comment]
Hey @woodruffw ! I have a PR almost ready aiming this improvement, would you mind assigning this issue to me? :)

**Comment by woodruffw** [comment]
Sure, thanks @ubiratansoares! 

BTW you might want to look at https://github.com/zizmorcore/zizmor/issues/2140 for some context -- I haven't had time to push it forwards (so a PR would be welcome) but @rmuir left a useful comment there about edge cases.

**Comment by rmuir** [comment]
IMO the actionlint "workaround" is pretty practical in use. The tool isn't updated nearly as often as zizmor and often finds behind. 

But in configuation I'm able to add all the corner cases as allow-listed (kinda: i declare them as self-hosted which isn't true). e.g. i'm using the 26.04 preview runners in an open source project so i've got this as a workaround in actionlint.yml

```yaml
# not yet recognized as runners in actionlint's static list
# declare as self-hosted so it doesn't complain about invalid labels
self-hosted-runner:
  labels: ["ubuntu-26.04", "ubuntu-26.04-arm"]
```

A similar allow-list type approach to an audit here would keep it clean for the various corner-cases such as github-hosted-runners-in-preview, github-hosted-runners-with-private-azure-networking, github-hosted-runners-with-custom-images, etc etc. And it gives the user a workaround that eliminates a race with stuff like "preview" images too.

**Comment by ubiratansoares** [comment]
I missed https://github.com/zizmorcore/zizmor/pull/2140 for some reason, but my take was quite similar to yours without having a look (and was a good refresh/ramp-up with new things in the code base btw!). 

I'm still missing expanding some expressions and matching against gh-hosted runners to raise it. I did not think automating updates for a list of gh-hosted runners, I came up with a static list of runners myself. 

Do you see automation as a must have to get started? I had doubts about because we miss a realiable API to use as an authoritative source for available `runs-on` labels, the closest thing I could find is [endoflife.date](https://endoflife.date/github-actions-runner-images), which [has an API ](https://endoflife.date/github-actions-runner-images)but does not expose the labels (yet) neither support larger runner variants. 

It does expose _decommissioned_ Runners though, which helps preventing some FPs when checking Workflows using them (for whatever reasons). I included the same runners on my hand-made list to catch FPs

**Comment by woodruffw** [comment]
> Do you see automation as a must have to get started?

Nope, I think it's optional! I agree it's hard to find a reliable data source here; having a hand-curated version seems totally fine for now.

**Comment by woodruffw** [comment]
#2350
